libs/@local/harpc/net/src/session/test.rs

```diff
@@ -0,0 +1,426 @@
+/// remove as many variables as possible from RTT measurements.
+struct SimpleEchoService;
+
+impl SimpleEchoService {
+    async fn handle(
+        mut sink: TransactionSink,
+        mut stream: TransactionStream,
+    ) -> error_stack::Result<(), PingError> {
+        while let Some(received) = stream.next().await {
+            sink.send(Ok(received))
+                .await
+                .change_context(PingError::Sink)?;
+        }
+
+        if stream.is_incomplete() == Some(true) {
+            return Err(Report::new(PingError::IncompleteStream));
+        }
+
+        Ok(())
+    }
+
+    fn accept(sink: TransactionSink, stream: TransactionStream) {
+        tokio::spawn(async move {
+            if let Err(error) = Self::handle(sink, stream).await {
+                tracing::error!(?error, "failed to handle transaction");
+            }
+        });
+    }
+
+    fn spawn(mut server: ListenStream) {
+        tokio::spawn(async move {
+            while let Some(transaction) = server.next().await {
+                let (_, sink, stream) = transaction.into_parts();
+                Self::accept(sink, stream);
+            }
+        });
+    }
+}
+
+fn server(
+    transport_config: TransportConfig,
+    session_config: server::SessionConfig,
+    transport: impl Transport,
+) -> (server::SessionLayer<StringEncoder>, impl Drop) {
+    let cancel = CancellationToken::new();
+
+    let transport_layer = TransportLayer::start(transport_config, transport, cancel.clone())
+        .expect("failed to start transport layer");
+
+    let session_layer = server::SessionLayer::new(session_config, transport_layer, StringEncoder);
+
+    (session_layer, cancel.drop_guard())
+}
+
+fn client(
+    transport_config: TransportConfig,
+    session_config: client::SessionConfig,
+    transport: impl Transport,
+) -> (client::SessionLayer, impl Drop) {
+    let cancel = CancellationToken::new();
+
+    let transport_layer = TransportLayer::start(transport_config, transport, cancel.clone())
+        .expect("failed to start transport layer");
+
+    let session_layer = client::SessionLayer::new(session_config, transport_layer);
+
+    (session_layer, cancel.drop_guard())
+}
+
+async fn echo<T>(mut transport: impl FnMut() -> T + Send, address: Multiaddr)
+where
+    T: Transport,
+{
+    let (server, _server_guard) = server(
+        TransportConfig::default(),
+        server::SessionConfig::default(),
+        transport(),
+    );
+    let server_ipc = server.transport().ipc().clone();
+
+    let address = {
+        let server_stream = server
+            .listen(address)
+            .await
+            .expect("should be able to listen on TCP");
+
+        // Give the swarm some time to acquire the external address
+        // This is necessary for CI, as otherwise the tests are a bit flaky.
+        tokio::time::sleep(tokio::time::Duration::from_millis(50)).await;
```

Comment:
```suggestion
        // Give the swarm some time to acquire the external address
        // This is necessary for CI, as otherwise the tests are a bit flaky.
        // TODO: Implement waiting for server to be ready
        //   see https://linear.app/hash/issue/H-2837/[bilal]-implement-waiting-for-server-to-be-ready
        tokio::time::sleep(tokio::time::Duration::from_millis(50)).await;
```